plugin/pkg/admission/persistentvolumeclaim/default/admission.go

```diff
@@ -0,0 +1,172 @@
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return kclient.Extensions().StorageClasses().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return kclient.Extensions().StorageClasses().Watch(options)
+			},
+		},
+		&extensions.StorageClass{},
+		store,
+		0,
+	)
+
+	return &claimDefaulterPlugin{
+		Handler:   admission.NewHandler(admission.Create, admission.Update),
+		client:    kclient,
+		store:     store,
+		reflector: reflector,
+	}
+}
+
+func (a *claimDefaulterPlugin) Run() {
+	if a.stopChan == nil {
+		a.stopChan = make(chan struct{})
+	}
+	a.reflector.RunUntil(a.stopChan)
+}
+func (a *claimDefaulterPlugin) Stop() {
+	if a.stopChan != nil {
+		close(a.stopChan)
+		a.stopChan = nil
+	}
+}
+
+// This is a stand-in until we have a real field.  This string should be a const somewhere.
+const classAnnotation = "volume.beta.kubernetes.io/storage-class"
+
+// This indicates that a particular StorageClass nominates itself as the system default.
+const isDefaultAnnotation = "storage-class.beta.kubernetes.io/isDefaultClass"
+
+// Admit sets the default value of a PersistentVolumeClaim's storage class, in case the user did
+// not provide a value.
+//
+// 1.  Find available StorageClasses.
+// 2.  Figure which is the default
+// 3.  Write to the PVClaim
+func (c *claimDefaulterPlugin) Admit(a admission.Attributes) error {
+	if a.GetResource().GroupResource() != api.Resource("persistentvolumeclaims") {
+		return nil
+	}
+
+	if len(a.GetSubresource()) != 0 {
+		return nil
+	}
+
+	pvc, ok := a.GetObject().(*api.PersistentVolumeClaim)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	_, found := pvc.Annotations[classAnnotation]
+	if found {
+		// The use asked for a class.
+		return nil
+	}
+
+	glog.V(4).Infof("no storage class for claim %s (generate: %s)", pvc.Name, pvc.GenerateName)
+
+	def, err := getDefaultClass(c.store)
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+	if def == nil {
+		// No default class selected, do nothing about the PVC.
+		return nil
+	}
+
+	glog.V(4).Infof("defaulting storage class for claim %s (generate: %s) to %s", pvc.Name, pvc.GenerateName, def.Name)
+	if pvc.ObjectMeta.Annotations == nil {
+		pvc.ObjectMeta.Annotations = map[string]string{}
+	}
+	pvc.Annotations[classAnnotation] = def.Name
+	return nil
+}
+
+// getDefaultClass returns the default StorageClass from the store, or nil.
+func getDefaultClass(store cache.Store) (*extensions.StorageClass, error) {
+	defaultClasses := []*extensions.StorageClass{}
+	for _, c := range store.List() {
+		class, ok := c.(*extensions.StorageClass)
+		if !ok {
+			return nil, errors.NewInternalError(fmt.Errorf("error converting stored object to StorageClass: %v", c))
+		}
+		if class.Annotations[isDefaultAnnotation] == "true" {
```

Comment:
requiring per-zone classes is horrible, but it is fallout from inability to
reach a final design in this time window.  But it ONLY matters if the user
is asking for a PVC in a specific zone.

Do we expect most users to specify a zone for their claims or to let the
system spread them?  I expect the latter

On Fri, Aug 19, 2016 at 9:26 AM, Mark Turansky <notifications@github.com>
wrote:

> In plugin/pkg/admission/persistentvolumeclaim/default/admission.go
> <https://github.com/kubernetes/kubernetes/pull/30900#discussion_r75511207>
> :
>
> > +	if pvc.ObjectMeta.Annotations == nil {
> > +		pvc.ObjectMeta.Annotations = map[string]string{}
> > +	}
> > +	pvc.Annotations[classAnnotation] = def.Name
> > +	return nil
> > +}
> > +
> > +// getDefaultClass returns the default StorageClass from the store, or nil.
> > +func getDefaultClass(store cache.Store) (*extensions.StorageClass, error) {
> > +	defaultClasses := []*extensions.StorageClass{}
> > +	for _, c := range store.List() {
> > +		class, ok := c.(*extensions.StorageClass)
> > +		if !ok {
> > +			return nil, errors.NewInternalError(fmt.Errorf("error converting stored object to StorageClass: %v", c))
> > +		}
> > +		if class.Annotations[isDefaultAnnotation] == "true" {
>
> Let me rephrase ...
>
> I would agree with 1-2 classes for most setups *IF* most users are
> running small clusters in a single zone.
>
> For anyone running multi-zone -- and I don't know if this is the norm --
> you'll probably have the same 1-2 classes *per zone*.
>
> This specific thing is coming up right now for discussion downstream in
> OpenShift Online.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30900/files/0fb0c07b4cb9c07d30a0d4d27cb4e726db33705a#r75511207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVMhvZ4wSF_IUoVYDjeaMxUSN3sm4ks5qhdkwgaJpZM4JntWR>
> .
>